### ISSUE TITLE
Add pre-commit hook to enforce file mode permissions

### ISSUE THREAD

**Issue by sds** [brigade/overcommit#191]
It's common when working in a collaborative environment to have individuals who have not configured their editors to save files with appropriate default file permissions. Usually this manifests itself as source files being committed that have unnecessary execute permissions.

It would be nice to have a `FileMode` pre-commit hook that allows you to enforce certain permissions for files matching specified glob patterns, e.g.

```yaml
PreCommit:
  FileMode:
    enabled: true
    mode:
      0644: '**/*rb'
      0755: 'bin/*'
```

Some unknowns here are how to deal with overlapping globs (is the first match the one that is enforced?), and how this plays out on Windows (which IIRC all files are executable). Leaving this here for comments.

**Comment by jawshooah** [comment]
Not sure what to do here for Windows, but I think enforcing the first match makes sense. That said, won't the config be parsed into a hash? Is key order preserved here?

**Comment by sds** [comment]
Yes, Ruby uses an ordered hash implementation so that insertion order of keys is preserved. https://www.igvita.com/2009/02/04/ruby-19-internals-ordered-hash/

Digging a bit deeper, it looks like the only permissions git actually tracks is execute permissions (http://blog.lesc.se/2011/11/how-to-change-file-premissions-in-git.html). Your system's `umask` determines the default file permissions otherwise.

Thus a better name for such a hook is probably `ExecutePermissions`, since git doesn't actually control any other permissions.

This also presents an interesting question: since (I assume) configuring default file permissions for Windows is a pain, would it make sense for this hook to manually update the permissions based on your settings? For example, `git update-index --chmod=+x <file>` sets the execute permission in git's index, so we could run that on pre-commit and stage those changes before committing. This would be the first pre-commit hook to stage additional files on your behalf, but I think in this case it would make sense.

**Comment by jawshooah** [comment]
Could be useful, though we'd have to make it painfully clear that the hook actually modifies file permissions in the index, rather than just checking them.

**Comment by lencioni** [comment]
> would it make sense for this hook to manually update the permissions based on your settings?

I'm not a big fan of hooks making changes automatically. It seems safest to just tell people what is wrong and let them decide how to resolve it.

**Comment by sds** [comment]
Implemented in 526a3709e6e667eef3f328493401e6ec7d9bf309. This will not touch the index (since that's not yet possible with pre-commit hooks), and is not enabled by default.
